rir/src/compiler/opt/inline.cpp

```diff
@@ -200,14 +200,11 @@ class TheInliner {
                         }
                         if (ld) {
                             Value* a = arguments[ld->id];
-                            if (auto mk = MkArg::Cast(a)) {
+                            if (MkArg::Cast(a)) {
                                 // We need to cast from a promise to a lazy
                                 // value
-                                auto cast = new CastType(
-                                    a, RType::prom,
-                                    mk->isEager()
-                                        ? mk->eagerArg()->type.forced()
-                                        : ld->type);
+                                auto cast =
```

Comment:
@mhyee that was the bug you where seeing.

when we inline we do the following:
```
   prom  %1 = mkarg(...)
   val^   %2 = Cast(val^, %1)
```

we add a cast to convert promise-as-a-value to lazy-value. The former is a promise in value position, ie. we mean the promise itself. The later is a lazy value, ie. we mean the value we get out of the promise. That cast is needed, because normally the call instruction does this implicit "conversion".

Now I tried to be clever and already propagate the underlying type, if the argument was eager. But that caused the compiler to thing that the result of `Cast` is a value without promise wrapper, when in fact it has a promise wrapper, that needs to be removed.

Most of the time this does not cause any issues, since many operations just implicitly evaluate promises. But the `asbool_` instruction does not do so...